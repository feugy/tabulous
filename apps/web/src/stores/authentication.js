import { BehaviorSubject } from 'rxjs'
import { initGraphQLGlient, runMutation } from './graphql-client'
import * as graphQL from '../graphql'
import { makeLogger } from '../utils'

const logger = makeLogger('authentication')

const storageKey = 'player'

const current$ = new BehaviorSubject()

current$.subscribe(player => {
  if (player) {
    logger.debug({ player }, `saving session`)
    sessionStorage.setItem(storageKey, JSON.stringify(player))
    initGraphQLGlient(player)
  } else if (player === null) {
    logger.debug(`clearing session storage`)
    sessionStorage.clear()
    initGraphQLGlient()
  }
})

/**
 * Emits currently authenticated player
 * @type {Observable<object>}
 */
export const currentPlayer = current$.asObservable()

/**
 * Recovers previous session by reusing data from session storage.
 * @async
 * @returns {object|null} the authenticated player in case of success, or null
 */
export async function recoverSession() {
  const playerData = sessionStorage.getItem(storageKey)
  if (playerData) {
    let player = null
    try {
      logger.debug({ playerData }, `recovering previous session`)
      player = await logIn(JSON.parse(playerData).username)
    } catch (error) {
      logger.warn(
        { error, playerData },
        `failed to recover session: ${error.message}`
      )
      await logOut()
    }
    logger.debug({ player: current$.value }, `session recovery complete`)
    return player
  }
}

/**
 * Logs a player in.
 * @async
 * @param {string} username - username of the authenticating player
 * @param {string} password - clear password
 * @returns {object} the authenticated player object
 */
export async function logIn(username, password) {
  initGraphQLGlient()
  const player = await runMutation(graphQL.logIn, { username, password })
  logger.info(player, `authenticating ${username}`)
  current$.next(player)
  return player
}

/**
 * Logs current player out.
 * @async
 */
export async function logOut() {
  current$.next(null)
}
