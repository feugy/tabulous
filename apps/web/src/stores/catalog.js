import { runQuery } from './graphql-client'
import * as graphQL from '../graphql'
import { makeLogger } from '../utils'

const logger = makeLogger('players')

/**
 * @typedef {object} CatalogItem a catalog item
 * @property {string} name - item unique name.
 */

/**
 * List all catalog items.
 * @returns {Promise<CatalogItem[]>} a list of catalog items.
 */
export async function listCatalog() {
  logger.info('list catalog')
  return runQuery(graphQL.listCatalog)
}
