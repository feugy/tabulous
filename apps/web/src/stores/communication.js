import Peer from 'peerjs'
import { BehaviorSubject, Subject } from 'rxjs'
import { makeLogger } from '../utils'

const logger = makeLogger('communication')

const storageKey = 'peers'

let peer

const connections = new Map()
const lastMessageSent$ = new Subject()
const lastMessageReceived$ = new Subject()
const connected$ = new BehaviorSubject()
const lastConnected$ = new Subject()

function removeConnection(connection) {
  logger.info({ connection }, `connection to ${connection.peer} lost!`)
  connections.delete(connection.peer)
  connected$.next([...(peer ? [peer.id] : []), ...connections.keys()])
}

function setupConnection(connection) {
  connection.on('open', () => {
    const peers = connected$.value
    logger.info(
      { peers, connection },
      `connection established with ${connection.peer}, sending peers`
    )
    connection.send({ peers })
  })
  connection.on('error', error =>
    logger.warn(
      { connection, error },
      `${connection.peer} encounter an error: ${error.message}`
    )
  )
  connection.on('data', data => {
    logger.debug({ data, connection }, `data from ${connection.peer}`)
    lastMessageReceived$.next({ data, from: connection.peer })
  })
  connection.on('close', () => {
    logger.info({ connection }, `connection to ${connection.peer} closed`)
    removeConnection(connection)
  })
  connections.set(connection.peer, connection)
  connected$.next([peer.id, ...connections.keys()])
}

connected$.subscribe(peers => {
  if (peers) {
    logger.debug({ peers }, `persisting current peers`)
    sessionStorage.setItem(storageKey, JSON.stringify(peers))
  }
})

export const lastMessageSent = lastMessageSent$.asObservable()

export const lastMessageReceived = lastMessageReceived$.asObservable()

export const connected = connected$.asObservable()

export const lastConnected = lastConnected$.asObservable()

export async function initCommunication(id) {
  logger.info({ id }, `initializing communication`)
  const previousPeers = sessionStorage.getItem(storageKey)
  for (const connection of [...connections]) {
    removeConnection(connection)
  }

  peer = new Peer(id, { debug: 3 })
  peer.on('connection', connection => {
    setupConnection(connection)
    connection.on('open', () => lastConnected$.next(connection.peer))
  })
  connected$.next([peer.id])
  if (previousPeers) {
    logger.debug({ previousPeers }, `try to reconnect with previous peers...`)
    try {
      const [previousId, ...peers] = JSON.parse(previousPeers)
      if (previousId === id) {
        for (const peer of peers) {
          console.log(connections.has(peer), peer)
          await connectWith(peer, false)
        }
      }
    } catch (error) {
      logger.warn(
        { error, previousPeers },
        `can not reconnect with previous peers: ${error.message}`
      )
    }
  }
}

export async function connectWith(id, waitForPeers = true) {
  if (!peer || connections.has(id)) return
  setupConnection(
    await new Promise((resolve, reject) => {
      const connection = peer.connect(id)
      logger.info(
        { peer: id, waitForPeers },
        `establishing connection with peer ${id}`
      )

      connection.once('open', () => {
        logger.info(
          { connection },
          `connection established with ${connection.peer}`
        )
        if (waitForPeers) {
          connection.on('data', receivePeers)
        } else {
          resolve(connection)
        }
      })

      connection.once('error', error => {
        logger.warn(
          { error, peer: id },
          `failed to connect with ${id}: ${error.message}`
        )
        reject(error)
      })

      function receivePeers(data) {
        if (data?.peers) {
          connection.removeListener('data', receivePeers)
          logger.info(
            { ...data, connection },
            `receiving peers from ${connection.peer}`
          )
          for (const peer of data.peers) {
            if (connections.has(peer)) {
              connectWith(peer, false)
            }
          }
          resolve(connection)
        }
      }
    })
  )
}

export function send(data, to = null) {
  if (peer) {
    const closed = []
    lastMessageSent$.next({ data, from: peer.id })
    if (to && !connections.has(to)) {
      return
    }
    const destination = to ? new Map([[to, connections.get(to)]]) : connections
    for (const [, connection] of destination) {
      if (connection.open) {
        logger.debug({ data, connection }, `sending data to ${connection.peer}`)
        connection.send(data)
      } else {
        closed.push(connection)
      }
    }
    if (closed.length) {
      for (const connection of closed) {
        removeConnection(connection)
      }
    }
  }
}
