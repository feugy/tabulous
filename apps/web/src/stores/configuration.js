import { BehaviorSubject } from 'rxjs'

/**
 * @typedef {object} Flags configuration flags
 * @property {boolean} useGithubProvider enables "Connect with Github" button.
 * @property {boolean} useGoogleProvider enables "Connect with Google" button.
 */
const flags$ = new BehaviorSubject({
  useGithubProvider: import.meta.env.WEB_USE_GITHUB_PROVIDER,
  useGoogleProvider: import.meta.env.WEB_USE_GOOGLE_PROVIDER
})

/**
 * @type {import('rxjs').Observable<Flags>}
 */
export const flags = flags$.asObservable()
