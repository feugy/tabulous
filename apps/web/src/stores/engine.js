import { BehaviorSubject, Subject } from 'rxjs'
import {
  controlManager,
  dragManager,
  multiSelectionManager
} from '../3d/managers'
import * as utils from '../3d/utils'

const engine$ = new BehaviorSubject(null)
const fps$ = new BehaviorSubject(0)
// event streams
const pointerOut$ = new Subject()
const pointerOver$ = new Subject()
const selectionActive$ = new Subject()
const selectionReset$ = new Subject()
const dragStart$ = new Subject()
const drag$ = new Subject()
const dragEnd$ = new Subject()
const action$ = new Subject()
const pointer$ = new Subject()

const mapping = [
  { observable: multiSelectionManager.onOverObservable, subject: pointerOver$ },
  { observable: multiSelectionManager.onOutObservable, subject: pointerOut$ },
  {
    observable: multiSelectionManager.onSelectionActiveObservable,
    subject: selectionActive$
  },
  {
    observable: multiSelectionManager.onSelectionResetObservable,
    subject: selectionReset$
  },
  { observable: dragManager.onDragStartObservable, subject: dragStart$ },
  { observable: dragManager.onDragObservable, subject: drag$ },
  { observable: dragManager.onDragEndObservable, subject: dragEnd$ },
  { observable: controlManager.onActionObservable, subject: action$ },
  { observable: controlManager.onPointerObservable, subject: pointer$ }
]

export const engine = engine$.asObservable()
export const fps = fps$.asObservable()
export const pointerOver = pointerOver$.asObservable()
export const pointerOut = pointerOut$.asObservable()
export const selectionActive = selectionActive$.asObservable()
export const selectionReset = selectionReset$.asObservable()
export const dragStart = dragStart$.asObservable()
export const drag = drag$.asObservable()
export const dragEnd = dragEnd$.asObservable()
export const action = action$.asObservable()
export const pointer = pointer$.asObservable()

export function initEngine(engine) {
  engine.onEndFrameObservable.add(() => fps$.next(engine.getFps().toFixed()))
  // expose Babylon observables as RX subjects
  for (const { observable, subject } of mapping) {
    mapping.observer = observable.add(subject.next.bind(subject))
  }
  engine$.next(engine)
}

export const applyAction = controlManager.apply.bind(controlManager)

export const movePeerPointer = controlManager.movePeerPointer.bind(
  controlManager
)

export function serializeScene() {
  return engine$.value ? utils.serializeScene(engine$.value.scenes[0]) : {}
}

export function loadScene(descriptor) {
  if (engine$.value) {
    utils.loadScene(engine$.value, engine$.value.scenes[0], descriptor)
  }
}
