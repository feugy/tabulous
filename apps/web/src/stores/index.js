export * from './catalog'
export * from './configuration'
export * from './discussion'
export * from './fullscreen'
export * from './game-engine'
export * from './game-manager'
export * from './graphql-client'
export * from './indicators'
export * from './notifications'
export * from './peer-channels'
export * from './players'
export * from './stream'
export * from './toaster'
