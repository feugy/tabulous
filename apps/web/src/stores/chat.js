import { merge } from 'rxjs'
import { filter, scan } from 'rxjs/operators'
import { lastMessageReceived, lastMessageSent, send } from './communication'

export const chat = merge(lastMessageSent, lastMessageReceived).pipe(
  filter(data => data.message),
  scan((thread, message) => [...thread, message], [])
)

export function sendMessage(message) {
  send({ message })
}
