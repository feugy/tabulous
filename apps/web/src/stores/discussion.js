import { BehaviorSubject, merge } from 'rxjs'
import { filter, scan } from 'rxjs/operators'
import { lastMessageReceived, lastMessageSent, send } from './communication'

const resetSymbol = Symbol('reset-chat')

// initialize to ensure an empty thread, instead of undefined
const reset$ = new BehaviorSubject({ message: resetSymbol })

export const thread = merge(lastMessageSent, lastMessageReceived, reset$).pipe(
  filter(data => data?.message),
  scan(
    (thread, data) => (data.message === resetSymbol ? [] : [...thread, data]),
    []
  )
)

export function sendToThread(message) {
  send({ message })
}

export function clearThread() {
  reset$.next({ message: resetSymbol })
}
