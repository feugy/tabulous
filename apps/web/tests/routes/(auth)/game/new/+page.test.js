import { faker } from '@faker-js/faker'
import NewGamePage from '@src/routes/(auth)/game/new/+page.svelte'
import { createGame } from '@src/stores'
import { render } from '@testing-library/svelte'
import { sleep } from '@tests/test-utils'
import { get } from 'svelte/store'
import html from 'svelte-htm'
import { beforeEach, describe, expect, it, vi } from 'vitest'

import { goto } from '$app/navigation'
import { page } from '$app/stores'

vi.mock('@src/stores')

describe('/game/new route', () => {
  const game = {
    id: faker.datatype.uuid(),
    name: faker.commerce.productMaterial()
  }

  beforeEach(vi.resetAllMocks)

  it('displays error when too many games where created', async () => {
    get(page).url.searchParams.set('name', game.name)
    createGame.mockRejectedValueOnce(
      new Error('You own 6 games, you can not create more')
    )
    const { container } = render(html`<${NewGamePage} />`)
    await sleep()
    expect(createGame).toHaveBeenCalledWith(game.name)
    expect(container).toMatchSnapshot()
    expect(goto).not.toHaveBeenCalled()
  })

  it('displays error game is restricted', async () => {
    get(page).url.searchParams.set('name', game.name)
    createGame.mockRejectedValueOnce(new Error('Access to game is restricted'))
    const { container } = render(html`<${NewGamePage} />`)
    await sleep()
    expect(createGame).toHaveBeenCalledWith(game.name)
    expect(container).toMatchSnapshot()
    expect(goto).not.toHaveBeenCalled()
  })

  it('redirects to game on success', async () => {
    get(page).url.searchParams.set('name', game.name)
    createGame.mockResolvedValueOnce(game.id)
    render(html`<${NewGamePage} />`)
    await sleep()
    expect(createGame).toHaveBeenCalledWith(game.name)
    expect(goto).toHaveBeenCalledWith(game.id, { replaceState: true })
    expect(goto).toHaveBeenCalledTimes(1)
  })

  it('redirects to home without game name', async () => {
    get(page).url.searchParams.delete('name')
    render(html`<${NewGamePage} />`)
    await sleep()
    expect(createGame).not.toHaveBeenCalled()
    expect(goto).toHaveBeenCalledWith('/home')
    expect(goto).toHaveBeenCalledTimes(1)
  })
})
