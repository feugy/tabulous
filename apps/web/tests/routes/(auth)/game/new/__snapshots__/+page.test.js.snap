// Vitest Snapshot v1

exports[`/game/new route > displays error game is restricted 1`] = `
<body>
  <div>
     
    <main
      class="s-1rlJbZ_1rgK_"
    >
      <header
        class="s-IEV6mBVyuaCJ"
      >
        <nav
          class="s-IEV6mBVyuaCJ"
        >
          <span
            class="s-IEV6mBVyuaCJ"
          >
            <ol
              class="s-x5nEMCoiqeIy"
            >
              <li
                class="s-x5nEMCoiqeIy"
              >
                <a
                  class="s-x5nEMCoiqeIy"
                  href="/"
                >
                  Accueil
                </a>
              </li>
               
              <li
                class="s-x5nEMCoiqeIy"
              >
                &gt;
              </li>
              
              <li
                class="s-x5nEMCoiqeIy"
              >
                <a
                  class="s-x5nEMCoiqeIy"
                >
                  Nouvelle partie
                </a>
              </li>
               
              
            </ol>
            
            <!--&lt;Breadcrumb&gt;-->
          </span>
           
          <button
            class="secondary s-NMyTiX1zi6rz"
          >
            <span
              class="material-icons s-NMyTiX1zi6rz"
            >
              account_circle
            </span>
            
            
            
          </button>
          <!--&lt;Button&gt;-->
        </nav>
         
        <div
          class="s-IEV6mBVyuaCJ"
        />
      </header>
      <!--&lt;Header&gt;-->
       
      <section
        class="s-1rlJbZ_1rgK_"
      >
        <p
          class="s-1rlJbZ_1rgK_"
        >
          Ce jeu n'est pas accessible.
        </p>
         
        <a
          class="s-1rlJbZ_1rgK_"
          href="/home"
        >
          <button
            class="s-NMyTiX1zi6rz"
          >
            <span
              class="material-icons s-NMyTiX1zi6rz"
            >
              arrow_back
            </span>
            
            <span
              class="text s-NMyTiX1zi6rz"
            >
              Accueil
            </span>
            
            
          </button>
          <!--&lt;Button&gt;-->
        </a>
      </section>
    </main>
    <!--&lt;+page&gt;-->
  </div>
</body>
`;

exports[`/game/new route > displays error when too many games where created 1`] = `
<body>
  <div>
     
    <main
      class="s-1rlJbZ_1rgK_"
    >
      <header
        class="s-IEV6mBVyuaCJ"
      >
        <nav
          class="s-IEV6mBVyuaCJ"
        >
          <span
            class="s-IEV6mBVyuaCJ"
          >
            <ol
              class="s-x5nEMCoiqeIy"
            >
              <li
                class="s-x5nEMCoiqeIy"
              >
                <a
                  class="s-x5nEMCoiqeIy"
                  href="/"
                >
                  Accueil
                </a>
              </li>
               
              <li
                class="s-x5nEMCoiqeIy"
              >
                &gt;
              </li>
              
              <li
                class="s-x5nEMCoiqeIy"
              >
                <a
                  class="s-x5nEMCoiqeIy"
                >
                  Nouvelle partie
                </a>
              </li>
               
              
            </ol>
            
            <!--&lt;Breadcrumb&gt;-->
          </span>
           
          <button
            class="secondary s-NMyTiX1zi6rz"
          >
            <span
              class="material-icons s-NMyTiX1zi6rz"
            >
              account_circle
            </span>
            
            
            
          </button>
          <!--&lt;Button&gt;-->
        </nav>
         
        <div
          class="s-IEV6mBVyuaCJ"
        />
      </header>
      <!--&lt;Header&gt;-->
       
      <section
        class="s-1rlJbZ_1rgK_"
      >
        <p
          class="s-1rlJbZ_1rgK_"
        >
          Vous avez déjà 6 parties en cours, vous ne pouvez pas en créer davantage.
        </p>
         
        <a
          class="s-1rlJbZ_1rgK_"
          href="/home"
        >
          <button
            class="s-NMyTiX1zi6rz"
          >
            <span
              class="material-icons s-NMyTiX1zi6rz"
            >
              arrow_back
            </span>
            
            <span
              class="text s-NMyTiX1zi6rz"
            >
              Accueil
            </span>
            
            
          </button>
          <!--&lt;Button&gt;-->
        </a>
      </section>
    </main>
    <!--&lt;+page&gt;-->
  </div>
</body>
`;
